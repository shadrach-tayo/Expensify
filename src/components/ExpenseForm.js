import React from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super();
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      focused: false,
      error: ''
    };
  }

  onDescriptionchange = e => {
    this.setState({ description: e.target.value });
  };

  onNoteChange = e => {
    this.setState({ note: e.target.value });
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({ amount: e.target.value });
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      focused
    }));
  };

  onSubmit = e => {
    e.preventDefault();
    const { description, amount, note, createdAt } = this.state;
    if (!description || !amount) {
      this.setState({ error: 'please fill in the description and the amount' });
    } else {
      this.setState({ error: '' });
      this.props.onSubmit({
        amount: parseInt(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        description,
        note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error ? <p>{this.state.error}</p> : ''}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionchange}
          />
          <input
            type="number"
            value={this.state.amount}
            placeholder="amount"
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add expense note (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
