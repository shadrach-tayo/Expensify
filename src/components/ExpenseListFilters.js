import React, { useState } from "react";
import { DateRangePicker } from "react-dates";
import { useDispatch, useSelector } from "react-redux";
// import { setTextFilter } from "../actions/filters";
// import {
//   sortByAmount,
//   sortByDate,
//   setStartDate,
//   setEndDate,
// } from "../actions/filters";
import { getFilters } from "../selectors";
import {
  setEndDate,
  setStartDate,
  setText,
  sortByAmount,
  sortByDate,
} from "../slices/filters";

function ExpenseListFilters(props) {
  // state
  const [focused, setFocused] = useState(null);

  // selectors
  const { text, startDate, endDate, sortBy } = useSelector(getFilters);

  // dispatcher
  const dispatch = useDispatch();
  const _sortByAmount = () => dispatch(sortByAmount());
  const _setTextFilter = (value) => dispatch(setText(value));
  const _sortByDate = () => dispatch(sortByDate());
  const _setStartDate = (date) => dispatch(setStartDate(date));
  const _setEndDate = (date) => dispatch(setEndDate(date));


  const onDatesChange = ({ startDate, endDate }) => {
    _setStartDate(startDate);
    _setEndDate(endDate);
  };

  const onFocusChange = (focused) => {
    setFocused(focused);
  };

  const onTextChange = (e) => _setTextFilter(e.target.value);

  const onSortByChange = (e) => {
    if (e.target.value === "date") _sortByDate();
    else _sortByAmount();
  };

  return (
    <div className="container">
      <div className="input-group">
        <div className="input-group__item">
          <input
            type="text"
            className="text-input"
            defaultValue={text}
            onChange={onTextChange}
            placeholder="Search for Expenses"
          />
        </div>
        <div className="input-group__item">
          <select className="select" value={sortBy} onChange={onSortByChange}>
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>
        </div>
        <div className="input-group__item">
          <DateRangePicker
            startDate={startDate}
            endDate={endDate}
            onDatesChange={onDatesChange}
            focusedInput={focused}
            onFocusChange={onFocusChange}
            showClearDates={true}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
        </div>
      </div>
    </div>
  );
}

// const mapStateToProps = (state) => {
//   return {
//     filters: state.filters,
//   };
// };

// const mapDispatchToProps = (dispatch) => ({
//   setTextFilter: (text) => dispatch(setTextFilter(text)),
//   sortByDate: () => dispatch(sortByDate()),
//   sortByAmount: () => dispatch(sortByAmount()),
//   setStartDate: (startDate) => dispatch(setStartDate(startDate)),
//   setEndDate: (endDate) => dispatch(setEndDate(endDate)),
// });

export default ExpenseListFilters;
