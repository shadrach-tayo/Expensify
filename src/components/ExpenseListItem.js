import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({
  dispatch,
  id,
  description,
  amount,
  createdAt,
  ...props
}) => (
  <div>
    <Link to={`/edit/${id}`}>{description}</Link>
    <p>{numeral(amount / 100).format('$0,0.00')}</p>
    <p>{moment(createdAt).format('Do MMMM YYYY')}</p>
  </div>
);

export default ExpenseListItem;
