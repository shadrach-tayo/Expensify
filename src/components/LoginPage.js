import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => {
  return (
    <div>
      <p>This is the login page</p>
      <button onClick={startLogin}>Login</button>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(
  undefined,
  mapDispatchToProps
)(LoginPage);
