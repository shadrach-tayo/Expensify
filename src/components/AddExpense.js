import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpense extends React.Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h3>Add Expense</h3>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpense);
