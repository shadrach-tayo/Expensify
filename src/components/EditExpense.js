import React from 'react';
import { connect } from 'react-redux';
import ReactModal from 'react-modal';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

const modalStyle = {
  overlay: {
    backgroundColor: 'rgba(0, 0, 0, 0.8)'
  },
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
    display: 'flex',
    flexDirection: 'column',
    justifySelf: 'center',
    alignSelf: 'center',
    alignItems: 'center',
    padding: '10px 20px'
  }
};

ReactModal.setAppElement(document.getElementById('app'));

export class EditExpense extends React.Component {
  state = {
    showModal: false
  };
  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemove = () => {
    this.setState({ showModal: false });
    let id = this.props.expense.id;
    this.props.startRemoveExpense({ id });
    this.props.history.push('/');
  };

  confirmDelete = () => {
    this.setState({ showModal: true });
  };

  handleCloseModal = () => {
    this.setState({ showModal: false });
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="container">
            <h1 className="page-title">Edit Expense</h1>
          </div>
        </div>
        <div className="container">
          <ReactModal
            aria={{ label: 'Confirm expense delete modal' }}
            isOpen={this.state.showModal}
            shouldFocusAfterRender={true}
            shouldCloseOnEsc={true}
            shouldCloseOnOverlayClick={true}
            onRequestClose={this.handleCloseModal}
            style={modalStyle}
          >
            <p>
              <strong>{this.props.expense.description}</strong> will be
              permanently deleted
            </p>
            <button className="btn btn--delete" onClick={this.onRemove}>
              delete
            </button>
          </ReactModal>
          <ExpenseForm onSubmit={this.onSubmit} expense={this.props.expense} />
          <div className="container">
            <div className="input-group">
              <button
                className="btn btn--delete"
                onClick={this.confirmDelete}
                style={{ justifySelf: 'right' }}
              >
                Delete item
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: data => dispatch(startRemoveExpense(data))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense);
