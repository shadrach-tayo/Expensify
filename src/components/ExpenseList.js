import React from "react";
import { useSelector } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpense from "../selectors/expenses";
import { getExpenses, getFilters } from "../selectors";

export const ExpenseList = (props) => {
  const { expenses } = useSelector(getExpenses);
  const filters = useSelector(getFilters);

  const visibleExpenses = selectExpense(expenses, filters);
  console.log("visible ", visibleExpenses);

  return (
    <div className="container">
      <div className="list-header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop">Expense</div>
        <div className="show-for-desktop">Amount</div>
      </div>
      <div className="list-body">
        {visibleExpenses.length === 0 ? (
          <div className="list-item list-item__message">
            <span className="">No expenses</span>
          </div>
        ) : (
          visibleExpenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />;
          })
        )}
      </div>
    </div>
  );
};

// const mapStateToProps = (state) => {
//   return {
//     expenses: selectExpense(state.expenses, state.filters),
//   };
// };

export default ExpenseList;
