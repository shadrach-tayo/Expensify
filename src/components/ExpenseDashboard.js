import React, { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getExpenses } from "../selectors";
import { setExpense } from "../slices/expenses";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => {
  const dispatch = useDispatch();
  const _loadExpenses = useCallback(() => dispatch(setExpense(), [dispatch]));

  const { expenses, loading } = useSelector(getExpenses);

  useEffect(() => {
    if(loading) return;
    _loadExpenses();
    return () => {};
  }, []);


  return (
    <div>
      <ExpensesSummary />
      <ExpenseListFilters />
      <ExpenseList />
    </div>
  );
};

export default ExpenseDashboardPage;
