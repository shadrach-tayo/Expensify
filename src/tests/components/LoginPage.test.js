import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { LoginPage } from '../../components/LoginPage';

let wrapper, startLogin;
beforeEach(() => {
  startLogin = jest.fn();
  wrapper = shallow(<LoginPage startLogin={startLogin} />);
});

// react test renderer
test('should render LoginPage correctly', () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should call startLogin on button click', () => {
  wrapper.find('button').simulate('click');
  expect(startLogin).toHaveBeenCalled();
});
