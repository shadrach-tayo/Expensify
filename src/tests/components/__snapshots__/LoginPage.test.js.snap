// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LoginPage correctly 1`] = `
<div>
  <p>
    This is the login page
  </p>
  <button
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
